File: core/src/ru/myitschool/dcrawler/entities/ai/pathfinding/NodePath.java
package ru.myitschool.dcrawler.entities.ai.pathfinding;

import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.entities.ai.pathfinding.graph.Node;
import ru.myitschool.dcrawler.entities.ai.pathfinding.graph.NodeConnection;
import ru.myitschool.dcrawler.utils.AdvancedArray;

import java.util.Iterator;

/**
 * Created by Voyager on 01.05.2017.
 */
public class NodePath implements Comparable<NodePath> {

    private AdvancedArray<Node> nodes;
    private AdvancedArray<NodeConnection> connections;
    private int cost;

    private NodePath(Node startNode, Array<NodeConnection> connections, int cost) {
        AdvancedArray<Node> newNodes = new AdvancedArray<>();
        newNodes.add(startNode);
        for (int i = 0; i < connections.size; i++) {
            newNodes.add(connections.get(i).getToNode());
        }
        AdvancedArray<NodeConnection> newConnections = new AdvancedArray<>();
        for (int i = 0; i < connections.size; i++) {
            newConnections.add(connections.get(i));
        }
        this.nodes = newNodes;
        this.connections = newConnections;
        this.cost = cost;
    }

    public NodePath(Node startNode) {
        nodes = new AdvancedArray<>();
        nodes.add(startNode);
        connections = new AdvancedArray<>();
    }

    public int getNodeCount() {
        return nodes.size;
    }

    public Node get(int index) {
        return nodes.get(index);
    }

    public Node getLast(){
        return nodes.get(nodes.size - 1);
    }

    public boolean add(NodeConnection connection) {
        Node lastNode = nodes.getLast();
        if (connection.getFromNode().equals(lastNode)) {
            nodes.add(connection.getToNode());
            connections.add(connection);
            cost += connection.getCost();
            return true;
        }
        return false;
    }

    public void clear() {
        nodes.clear();
    }

    public void reverse() {
        nodes.reverse();
    }

    public Iterator<Node> iterator() {
        return null;
    }

    public Array<Node> getNodes() {
        return nodes;
    }

    public NodePath clone() {
        return new NodePath(nodes.get(0), connections, cost);
    }

    @Override
    public int compareTo(NodePath o) {
        return o.cost - this.cost;
    }

    @Override
    public String toString() {
        return cost + "";
    }

    public void remove(int index){
        nodes.removeIndex(index);
    }

    public int getCost() {
        return cost;
    }

    public NodePath cut(int limit){
        AdvancedArray<Node> nodes = new AdvancedArray<>();
        int cost = 0;
        for (int i = 0; i < limit; i++) {
            nodes.add(this.nodes.get(i));
            if (i > 0) {
                cost += this.nodes.get(i - 1).connectionCost(nodes.get(i));
            }
        }
        this.cost= cost;
        this.nodes = nodes;
        return this;
    }

    public NodePath cutLast(){
        return cut(nodes.size - 1);
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/ai/pathfinding/NodePath.kt
package com.wolg_vlad.dcrawler.entities.ai.pathfinding

import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.entities.ai.pathfinding.graph.Node
import com.wolg_vlad.dcrawler.entities.ai.pathfinding.graph.NodeConnection
import com.wolg_vlad.dcrawler.utils.AdvancedArray

/**
 * Created by Voyager on 01.05.2017.
 */
class NodePath : Comparable<NodePath> {
    var nodes: AdvancedArray<Node>
    private set
    private var connections: AdvancedArray<NodeConnection>
    var cost = 0
        private set

    private constructor(startNode: Node, connections: Array<NodeConnection>, cost: Int) {
        val newNodes = AdvancedArray<Node>()
        newNodes.add(startNode)
        for (i in 0 until connections.size) {
            newNodes.add(connections[i].toNode)
        }
        val newConnections = AdvancedArray<NodeConnection>()
        for (i in 0 until connections.size) {
            newConnections.add(connections[i])
        }
        nodes = newNodes
        this.connections = newConnections
        this.cost = cost
    }

    constructor(startNode: Node) {
        nodes = AdvancedArray()
        nodes.add(startNode)
        connections = AdvancedArray()
    }

    val nodeCount: Int
        get() = nodes.size

    operator fun get(index: Int): Node {
        return nodes[index]
    }

    val last: Node
        get() = nodes[nodes.size - 1]

    fun add(connection: NodeConnection): Boolean {
        val lastNode = nodes.last
        if (connection.fromNode == lastNode) {
            nodes.add(connection.toNode)
            connections.add(connection)
            cost += connection.cost
            return true
        }
        return false
    }

    fun clear() {
        nodes.clear()
    }

    fun reverse() {
        nodes.reverse()
    }

    operator fun iterator(): Iterator<Node>? {
        return null
    }

    fun clone(): NodePath {
        return NodePath(nodes[0], connections, cost)
    }

    override fun compareTo(other: NodePath): Int {
        return cost - other.cost
    }

    override fun toString(): String {
        return cost.toString() + ""
    }

    fun remove(index: Int) {
        nodes.removeIndex(index)
    }

    fun cut(limit: Int): NodePath {
        val nodes = AdvancedArray<Node>()
        var cost = 0
        for (i in 0 until limit) {
            nodes.add(this.nodes[i])
            if (i > 0) {
                cost += this.nodes[i - 1]!!.connectionCost(nodes[i])
            }
        }
        this.cost = cost
        this.nodes = nodes
        return this
    }

    fun cutLast(): NodePath {
        return cut(nodes.size - 1)
    }
}